runtime/doc/map.txt

```diff
@@ -1430,6 +1430,145 @@ Possible values are (second column is the short name used in listing):
     -addr=other		  ?	other kind of range
 
 
+Incremental preview ~
+                                                  *:command-preview* {nvim-api}
+It's possible to allow commands to be incrementally previewed when
+'inccommand' is enabled by using the preview command flag. The preview
+flag may only be used through the Lua API (see |nvim_create_user_command()|).
+
+The preview callback must be a Lua function. When command preview is
+triggered, the preview callback will be called with 3 arguments, an `opts` table
+which is the same as the table passed to |nvim_create_user_command()| callbacks,
+the preview namespace number that's used to set highlights for the preview,
+and the preview buffer number that's used for the preview window when
+`inccommand=split`. Note that if `inccommand=nosplit`, the preview buffer number
+will not be provided.
+
+In order for command preview to be shown, these steps need to be followed:
+
+1. Modify the current buffer as required for the preview (see
+   |nvim_buf_set_text()| and |nvim_buf_set_lines()|).
+2. If preview buffer is provided, add necessary text to the preview buffer.
+3. Add required highlights to the current buffer. If preview buffer is
+   provided, add required highlights to the the preview buffer as well. All
+   highlights must be added to the preview namespace which is provided as an
+   argument to the preview callback (see |nvim_buf_add_highlight()| and
+   |nvim_buf_set_extmark()| for help on how to add highlights to a namespace).
+4. Return an integer which represents the preview type. The preview type
+   tells Nvim what kind of preview is being shown, which allows it to carry
+   out the appropriate tasks to show the preview. There preview type can
+   either be 0, 1 or 2. Here's what the return values mean:
+   0: No preview is shown.
+   1: Preview is shown but preview window isn't opened even if
+      `inccommand=split`.
+   2: Preview is shown and preview window is opened if `inccommand=split`.
+
+   Note that if `inccommand=nosplit`, there's no difference between a return
+   value of 1 and a return value of 2 from the preview callback.
+
+Once preview is over, Nvim will automatically revert any changes made to the
+current buffer and preview buffer during the preview. All highlights set in
+the preview namespace will also be cleared.
```

Comment:
this doesn't seem to happen (clearing the highlights)

Reply:
> this doesn't seem to happen (clearing the highlights)

`buf_clear()` clears everything in the preview namespace when the preview buffer is (re-)opened on next iteration of the preview, so it's basically as if the highlights were cleared

Reply:
But we also intend highlighting in current buffer, yes?

Reply:
> But we also intend highlighting in current buffer, yes?

yes, it does clear highlights for the main buffer